File: jvm/periph-java/src/main/java/it/uhde/periph/chips/power/Ina226Minimal.java
package it.uhde.periph.chips.power;

import it.uhde.periph.transport.Transport;

import java.io.IOException;

/**
 * INA226 — 16-bit current/power monitor with I²C interface (minimal driver).
 *
 * <p>Measures shunt voltage, bus voltage, current, and power. Calibration is
 * computed from the shunt resistance and maximum expected current; the result
 * is written to the Calibration register at construction time.
 *
 * <p>Default I²C address: 0x40 (A0=GND, A1=GND).
 *
 * <h2>Configuration defaults</h2>
 * <ul>
 *   <li>Mode: continuous shunt+bus (7)</li>
 *   <li>Bus voltage conversion time: 1.1 ms (4)</li>
 *   <li>Shunt voltage conversion time: 1.1 ms (4)</li>
 *   <li>Averaging: 1 sample (0)</li>
 * </ul>
 */
public class Ina226Minimal {

    // --- Register addresses ---
    protected static final int REG_CONFIG    = 0x00;
    protected static final int REG_SHUNT     = 0x01;
    protected static final int REG_BUS       = 0x02;
    protected static final int REG_POWER     = 0x03;
    protected static final int REG_CURRENT   = 0x04;
    protected static final int REG_CAL       = 0x05;
    protected static final int REG_MASK_EN   = 0x06;
    protected static final int REG_ALERT_LIM = 0x07;
    protected static final int REG_MFR_ID    = 0xFE;
    protected static final int REG_DIE_ID    = 0xFF;

    /** Default configuration: mode=7, VBUSCT=4, VSHCT=4, AVG=0 → 0x4127. */
    protected static final int DEFAULT_CONFIG = 0x4127;

    protected final Transport transport;
    protected final double currentLsb;
    protected final int    cal;
    /** Stored MODE bits (2:0) for wake(). Updated by configure() and shutdown(). */
    protected int lastMode = 7;

    /**
     * Construct the driver with default shunt (0.1 Ω) and max current (2.0 A).
     *
     * @param transport I²C transport bound to the INA226 device address
     * @throws IOException on I²C error
     */
    public Ina226Minimal(Transport transport) throws IOException {
        this(transport, 0.1, 2.0);
    }

    /**
     * Construct the driver.
     *
     * <p>Computes {@code Current_LSB = maxCurrent / 32768} and
     * {@code CAL = int(0.00512 / (Current_LSB × rShunt))}, then writes
     * the default configuration register and calibration register.
     *
     * @param transport  I²C transport bound to the INA226 device address
     * @param rShunt     shunt resistor value in Ω (e.g. 0.1)
     * @param maxCurrent maximum expected current in A (e.g. 2.0)
     * @throws IOException on I²C error
     */
    public Ina226Minimal(Transport transport, double rShunt, double maxCurrent) throws IOException {
        this.transport  = transport;
        this.currentLsb = maxCurrent / 32768.0;
        this.cal        = (int) (0.00512 / (currentLsb * rShunt));
        writeReg(REG_CONFIG, DEFAULT_CONFIG);
        writeReg(REG_CAL, cal);
    }

    /**
     * Read the bus voltage.
     *
     * @return bus voltage in V (1.25 mV LSB, unsigned 16-bit)
     * @throws IOException on I²C error
     */
    public double voltage() throws IOException {
        return readReg(REG_BUS) * 1.25e-3;
    }

    /**
     * Read the shunt voltage.
     *
     * @return shunt voltage in V (2.5 µV LSB, signed 16-bit)
     * @throws IOException on I²C error
     */
    public double shuntVoltage() throws IOException {
        return readRegSigned(REG_SHUNT) * 2.5e-6;
    }

    /**
     * Read the current through the shunt resistor.
     *
     * <p>Requires a valid Calibration register value (written in the constructor).
     *
     * @return current in A (signed, Current_LSB per bit)
     * @throws IOException on I²C error
     */
    public double current() throws IOException {
        return readRegSigned(REG_CURRENT) * currentLsb;
    }

    /**
     * Read the calculated power.
     *
     * <p>Power = 25 × Current_LSB × raw power register (unsigned).
     *
     * @return power in W
     * @throws IOException on I²C error
     */
    public double power() throws IOException {
        return readReg(REG_POWER) * 25.0 * currentLsb;
    }

    // ---- low-level helpers ----

    /**
     * Write a 16-bit value to a register (big-endian, register-pointer protocol).
     *
     * @param reg register address
     * @param val 16-bit value
     * @throws IOException on I²C error
     */
    protected void writeReg(int reg, int val) throws IOException {
        transport.write(new byte[]{
                (byte) reg,
                (byte) (val >> 8),
                (byte) (val & 0xFF)
        });
    }

    /**
     * Read an unsigned 16-bit value from a register.
     *
     * @param reg register address
     * @return unsigned 16-bit value (0–65535)
     * @throws IOException on I²C error
     */
    protected int readReg(int reg) throws IOException {
        byte[] b = transport.writeRead(new byte[]{(byte) reg}, 2);
        return ((b[0] & 0xFF) << 8) | (b[1] & 0xFF);
    }

    /**
     * Read a signed 16-bit value from a register.
     *
     * @param reg register address
     * @return signed 16-bit value (-32768–32767)
     * @throws IOException on I²C error
     */
    protected int readRegSigned(int reg) throws IOException {
        byte[] b = transport.writeRead(new byte[]{(byte) reg}, 2);
        return (short) (((b[0] & 0xFF) << 8) | (b[1] & 0xFF));
    }
}

File: jvm/periph-kotlin/src/main/kotlin/it/uhde/periph/chips/power/Ina226Minimal.kt
package it.uhde.periph.chips.power

import it.uhde.periph.transport.Transport

/**
 * INA226 — 16-bit current/power monitor with I²C interface (minimal driver).
 *
 * Measures shunt voltage, bus voltage, current, and power. Calibration is
 * computed from the shunt resistance and maximum expected current; the result
 * is written to the Calibration register at construction time.
 *
 * Default I²C address: 0x40 (A0=GND, A1=GND).
 *
 * ## Configuration defaults
 * - Mode: continuous shunt+bus (7)
 * - Bus voltage conversion time: 1.1 ms (4)
 * - Shunt voltage conversion time: 1.1 ms (4)
 * - Averaging: 1 sample (0)
 */
open class Ina226Minimal @JvmOverloads constructor(
    protected val transport: Transport,
    rShunt: Double = 0.1,
    maxCurrent: Double = 2.0
) {
    protected val currentLsb: Double = maxCurrent / 32768.0
    protected val cal: Int = (0.00512 / (currentLsb * rShunt)).toInt()

    /** Stored MODE bits (2:0) for [Ina226Full.wake]. Updated by configure() and shutdown(). */
    protected var lastMode: Int = 7

    companion object {
        // Register addresses
        const val REG_CONFIG    = 0x00
        const val REG_SHUNT     = 0x01
        const val REG_BUS       = 0x02
        const val REG_POWER     = 0x03
        const val REG_CURRENT   = 0x04
        const val REG_CAL       = 0x05
        const val REG_MASK_EN   = 0x06
        const val REG_ALERT_LIM = 0x07
        const val REG_MFR_ID    = 0xFE
        const val REG_DIE_ID    = 0xFF

        /** Default configuration: mode=7, VBUSCT=4, VSHCT=4, AVG=0 → 0x4127. */
        const val DEFAULT_CONFIG = 0x4127
    }

    init {
        writeReg(REG_CONFIG, DEFAULT_CONFIG)
        writeReg(REG_CAL, cal)
    }

    /**
     * Read the bus voltage.
     *
     * @return bus voltage in V (1.25 mV LSB, unsigned 16-bit)
     */
    fun voltage(): Double = readReg(REG_BUS) * 1.25e-3

    /**
     * Read the shunt voltage.
     *
     * @return shunt voltage in V (2.5 µV LSB, signed 16-bit)
     */
    fun shuntVoltage(): Double = readRegSigned(REG_SHUNT) * 2.5e-6

    /**
     * Read the current through the shunt resistor.
     *
     * Requires a valid Calibration register value (written in the constructor).
     *
     * @return current in A (signed, Current_LSB per bit)
     */
    fun current(): Double = readRegSigned(REG_CURRENT) * currentLsb

    /**
     * Read the calculated power.
     *
     * Power = 25 × Current_LSB × raw power register (unsigned).
     *
     * @return power in W
     */
    fun power(): Double = readReg(REG_POWER) * 25.0 * currentLsb

    // ---- low-level helpers ----

    /**
     * Write a 16-bit value to a register (big-endian, register-pointer protocol).
     *
     * @param reg register address
     * @param val 16-bit value
     */
    protected fun writeReg(reg: Int, `val`: Int) {
        transport.write(byteArrayOf(
            reg.toByte(),
            (`val` shr 8).toByte(),
            (`val` and 0xFF).toByte()
        ))
    }

    /**
     * Read an unsigned 16-bit value from a register.
     *
     * @param reg register address
     * @return unsigned 16-bit value (0–65535)
     */
    protected fun readReg(reg: Int): Int {
        val b = transport.writeRead(byteArrayOf(reg.toByte()), 2)
        return ((b[0].toInt() and 0xFF) shl 8) or (b[1].toInt() and 0xFF)
    }

    /**
     * Read a signed 16-bit value from a register.
     *
     * @param reg register address
     * @return signed 16-bit value (-32768–32767)
     */
    protected fun readRegSigned(reg: Int): Int {
        val b = transport.writeRead(byteArrayOf(reg.toByte()), 2)
        return (((b[0].toInt() and 0xFF) shl 8) or (b[1].toInt() and 0xFF)).toShort().toInt()
    }
}
